### PR TITLE
#164069219 Enable user reset password.

### PR DESCRIPTION
**Pull request by BagzieGracious**
#### What does this PR do?
- This PR enables a user to reset their password
#### Description of Task to be completed?
- Send a reset password link via email
- Enable a user to reset password
#### How should this be manually tested?
- Make sure PostgreSQL is installed and create a database.
- Clone the repo, cd into `Ah-backend-aquaman` and checkout the branch `ft-reset-password-164069219`
    -  pip install -r requirements.txt
    -  python manage.py makemmigrations
    -  python manage.py migrate
    -  python manage.py runserver
- Use these endpoints to reset password:
    -  Password Link
       `http://127.0.0.1:8000/api/users/reset-password/`
        Then check your email for reset password link
    -  Password Change
       `http://127.0.0.1:8000/api/users/reset-password/change/`
#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
- [#164069219](https://www.pivotaltracker.com/story/show/164069219)

#### Screen shots

* The user navigates to the `http://127.0.0.1:8000/api/users/reset-password/` and sends a valid email

Screenshot 2019-03-11 at 12 43 12

* The user goes to email and clicks on reset password link

Screenshot 2019-03-11 at 12 46 01

* The user follows that link and then he/she provides their password

Screenshot 2019-03-11 at 12 44 15

### CODE REVIEW

File: authors/apps/authentication/tests/test_reset_password.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        """Tests if a reset password link is sent successfully."""
+        self.client.post(self.registration_url, valid_user, format='json')
+        response = self.client.post(self.reset_password_url, {"email": "bagendadeogracious@gmail.com"}, format='json')
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data['message'], "Please check your email for the reset password link.")
+
+    def test_email_key_error(self):
+        """Tests if a user has a provided an email field."""
+        self.client.post(self.registration_url, valid_user, format='json')
+        response = self.client.post(self.reset_password_url, {"eml": "bagendadeogracious@gmail.com"}, format='json')
```
Comment (CryceTruly) [inline authors/apps/authentication/tests/test_reset_password.py]: Thanks for the good work @BagzieGracious,  However, you have this functionality in 4 places in one file, you could put in a variable.

File: authors/apps/authentication/tests/test_reset_password.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    def test_successful_password_change(self):
+        """Tests if a user has changed the password successfully."""
+        token = str((jwt.encode({"email": "bagendadeogracious@gmail.com"}, settings.SECRET_KEY)).decode('utf-8'))
+        self.client.post(self.registration_url, valid_user, format='json')
+        response = self.client.patch(self.change_password_url+"?token="+token, {"password": "bagenda1234"}, format='json')
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.data['message'], "you have reset your password successfully.")
+
+    def test_valid_token(self):
```
Comment (CryceTruly) [inline authors/apps/authentication/tests/test_reset_password.py]: Thanks for the good work @BagzieGracious , However, I think the function name and the docstring do not reflect what you are actually doing in the method.

File: authors/apps/authentication/views.py [hunk 3]
```diff
@@ -106,3 +107,45 @@ def get(self, request):
...
+                "reset-password/change",
+                str((jwt.encode({"email": request.data['email']}, settings.SECRET_KEY)).decode('utf-8')),
+                "Reset Password",
+                "requested for password reset.",
+                request.data['email']
+            ]
+            Utilities.email_renderer(message)
+            return Response({"message": "Please check your email for the reset password link."}, status.HTTP_200_OK)
+        except KeyError:
+            raise exceptions.ValidationError("Email is required in-order to reset password.")
```
Comment (CryceTruly) [inline authors/apps/authentication/views.py]: thanks, @BagzieGracious ,we should let all messages we send to the user be as descriptive and correct as possible, we could change ```in-order``` to ```in order```.

**Review by CryceTruly — CHANGES_REQUESTED** [review]
Thanks, @BagzieGracious, Great work you have here, just a few comments, if you could ensure you keep all lines of code to 79 or lesser characters. That's the pep8 guideline. Otherwise awesome feature.

**Review by wasibani-roy — CHANGES_REQUESTED** [review]
I think in your PR on how to test i think you should add the part where after requesting a review they should go check there email and then follow the link from the email as opposed to put there the backend route directly but good work @BagzieGracious

**Review by wasibani-roy — DISMISSED** [review]

**Review by CryceTruly — DISMISSED** [review]
lgtm

File: authors/apps/authentication/views.py (hunk 3 above)
Comment (BagzieGracious) [reply authors/apps/authentication/views.py]: i have worked on it

**Review by joelethan — DISMISSED** [review]
LGTM

**Review by Peace-Apple — DISMISSED** [review]
Good work deogracious

**Review by CryceTruly — APPROVED** [review]
LGTM

**Review by Peace-Apple — APPROVED** [review]
Good work deogracious, works fine.

**Review by wasibani-roy — APPROVED** [review]
Changes have been made good work @BagzieGracious

**Pull request merged**